app/jobs/tweet_fetching_job.rb

```diff
@@ -0,0 +1,117 @@
+
+    TweetFetchingJob.perform_later(@data_set, @user_id)
+  end
+
+  # These exceptions indicate that the job may succeed if tried again.
+  rescue_from(Twitter::Error::InternalServerError,
+              Twitter::Error::BadGateway,
+              Twitter::Error::ServiceUnavailable,
+              Twitter::Error::GatewayTimeout) do |exception|
+    TweetFetchingJob.set(wait: TweetFetchingJob.backoff)
+                    .perform_later(@data_set, @user_id)
+  end
+
+  def perform(data_set, user_id)
+    @data_set = data_set
+    @user_id = user_id
+
+    store_data(fetch_tweets)
+
+    do_completion_bookkeeping
+  end
+
+  def reschedule_at(current_time, attempts)
+    current_time + TweetFetchingJob.backoff
+  end
+
+  private
+
+  def do_completion_bookkeeping(successful=true)
+    @data_set.unauthorized << @user_id unless successful
+    @data_set.processed << @user_id
+    @data_set.save
+    @data_set.finalize_when_ready
+  end
+
+  def fetch_tweets
+    twitter_client.user_timeline(
+      @user_id.to_i,
+      count: Rails.application.config.tweets_per_user,
+      tweet_mode: 'extended'
+    )
+  end
+
+  def store_data(tweets)
+    tweets.each do |tweet|
+      es_client.create index: @data_set.index_name,
+                       type: '_doc',
+                       body: tweet.to_json
+    end
+  end
+
+  def twitter_client
+    @twitter_client ||= Twitter::REST::Client.new do |config|
+      config.consumer_key        = ENV['TWITTER_CONSUMER_KEY']
+      config.consumer_secret     = ENV['TWITTER_CONSUMER_SECRET']
+      config.access_token        = ENV['TWITTER_OAUTH_TOKEN']
+      config.access_token_secret = ENV['TWITTER_OAUTH_SECRET']
+    end
+  end
+
+  def es_client
+    @es_client ||= Elasticsearch::Client.new
+  end
+
+  def self.backoff
+    limit = Rails.configuration.rate_limit_limit * 0.95 # safety margin
+    window = Rails.configuration.rate_limit_window * 60 # convert to seconds
+
+    # This messy exponential:
+    # - sets backoff to 0 (i.e. execute task immediately) when we have fewer
+    # than half of our limit in the queue
+    # - sets the backoff to the entire window size when the enqueued is near
+    # (95% of) our limit
+    # - in between, biases toward the shorter end, via exponential decay
+    # This means that when cohorts are small, we'll grab all the data as quickly
+    # as possible. When they're medium-sized, we'll still be pretty fast. As
+    # the cohort size approaches the limit, the time to fetch all data will
+    # approach the window size. When the cohort size is near or above the limit,
+    # the time to fetch all data will exceed the window size, in order to avoid
+    # rate limiting.
+    # In addition, since this looks at all enqueued TweetFetchers and not just
+    # those belonging to a given DataSet, when we are fetching data for multiple
+    # DataSets at once, they will cooperate to avoid hitting the window.
+    [2*window - (2.0**(-2*enqueued/limit))*4*window, 0].max
```

Comment:
I haven't traced through how this works fully, so this might be moot, but depending solely on the number of jobs in the queue frightens me a little. Is it possible jobs can end up waiting longer because there are jobs *after* them in the queue? 

You're doing as exponential rather than linear so it's faster for small cohorts, correct? I think that's beneficial as long as the size of cohorts themselves is something like an Exponential. But most of the cohorts we'll be dealing with are n=1000, and an expo backoff is unboundedly bad relative to maximal throughput as cohorts grow to infinity. This might all be irrelevant given the limits, but how long will we end up waiting between i=997, 998, and 999? I think this expo needs to be capped or made linear or we store the n for the last RATE_LIMIT_WINDOW somewhere and have all requests call a `throttle` function.

Reply:
This is assuming that the queue will consist only of TweetFetchingJobs. We can enforce that if we need to in future using `queue_as` to specify the queue name. I'll comment it.

The exponential makes it fast for small cohorts, yes, and fairly fast for medium-sized ones, while stretching it out for long ones. In the 997-999 range, we're waiting 1-2 seconds between jobs.

It builds in a cap of about twice the rate-limiting window (here, about 30 minutes). But I actually think capping is a problem, now that I think about it -- if we had, say, a 10K cohort, we'd end up stacking most of that cohort at almost 30 minutes out, and then we'd end up rate-limiting there. 

I've just convinced myself we can't have all 3 of [cap on length of time, rate limit avoidance, arbitrary cohort sizes]. Right now we have the first two. I could do the second two instead, e.g. via a linear function (or better yet a relu), but I can't do all 3.

I can also periodically check the rate-limit-status endpoint and use data from there to reschedule everything, but I couldn't figure out a safe way to do that.